Feature/reservoir collect button

pedromcunha opened the pull request:
## Description

* Modified the CollectButton to use Reservoir whenever an address is available instead of linking out externally
* Added logic to fetch floor prices from Reservoir to display pricing on the CollectButton
* Enhance readme with required env variables for local development

## Deploy Notes

- [ ] Need to add environment variables
  - `NEXT_PUBLIC_RESERVOIR_API_URL=https://onchainsummer.xyz/api/reservoir`
  - `RESERVOIR_API_KEY=X` (Reservoir will sponsor an API key for onchainsummer, we'll coordinate with the BASE team to deliver the key)
  - `ALLOWED_API_DOMAINS=https://onchainsummer.xyz`

## Screenshots (if appropriate)

<img width="1156" alt="Screenshot 2023-08-16 at 5 24 14 PM" src="https://github.com/base-org/onchainsummer.xyz/assets/6723339/1812413e-df7e-4d72-abfd-c4441ea3f380">
<img width="575" alt="Screenshot 2023-08-16 at 5 22 38 PM" src="https://github.com/base-org/onchainsummer.xyz/assets/6723339/de4661e9-ee91-41d1-bb9e-a9e08b3241a2">

wilsoncusack commented:
exciting @pedromcunha ! could you rebase

pedromcunha commented:
> exciting @pedromcunha ! could you rebase

Alright, should now be rebased.

wilsoncusack commented:
thanks @pedromcunha! just aligning internally

wilsoncusack commented:
We plan to merge this! @lauchness @ximxim would be great to get a review

lauchness commented on src/utils/getCollections.ts:
```diff
@@ -0,0 +1,37 @@
+    contract: contracts,
+  })
+
+  const response = await fetch(url.href)
```
@ximxim should we be caching this using the next functionality?

pedromcunha replied:
I would only cache this for a max of 30s, just to keep that floor price fresh. Let me know if it makes sense, I can add it in.

lauchness commented:
This is pretty neat

Code LGTM - but I'm noticing the `CollectDialog` seems to be in an infinite loading state 
<img width="744" alt="Screen Shot 2023-08-17 at 5 46 10 PM" src="https://github.com/base-org/onchainsummer.xyz/assets/51837850/439c1b33-4b8f-46f2-8ee7-a593e61f9b7b">


@pedromcunha any ideas what's up there?

pedromcunha commented:
> This is pretty neat
> 
> Code LGTM - but I'm noticing the `CollectDialog` seems to be in an infinite loading state <img alt="Screen Shot 2023-08-17 at 5 46 10 PM" width="744" src="https://user-images.githubusercontent.com/51837850/261435548-439c1b33-4b8f-46f2-8ee7-a593e61f9b7b.png">
> 
> @pedromcunha any ideas what's up there?

Are you running this locally? Need to make sure you've set the required env variables I added above. If you have, do you mind checking the console for any errors?

pedromcunha commented:
> > This is pretty neat
> > Code LGTM - but I'm noticing the `CollectDialog` seems to be in an infinite loading state <img alt="Screen Shot 2023-08-17 at 5 46 10 PM" width="744" src="https://user-images.githubusercontent.com/51837850/261435548-439c1b33-4b8f-46f2-8ee7-a593e61f9b7b.png">
> > @pedromcunha any ideas what's up there?
> 
> Are you running this locally? Need to make sure you've set the required env variables I added above. If you have, do you mind checking the console for any errors?

Alright I believe I fixed this. I think you were on the wrong chain/not connected. Added some logic/ux to handle this. Thanks for catching it! Let me know if it works for you now with the latest version.

wilsoncusack commented on src/components/CollectButton/CollectButton.tsx:
this works, but is a weird hierarchy. We should have a button that can have various states: connect, mint, collect, link out, etc. Making the top level thing "CollectButton" is odd imo

pedromcunha replied:
Good point. I cleaned up the CollectButton and made a new ExternalLinkButton. Now all that CollectButton does is facilitate collecting (buying). ExternalLinkButton is used when the address is 0x (undefined) to link out to the partner website.

pedromcunha replied:
I decided to leave in the connection check/switching networks just because the UX of the modal does require these things to be checked. Hoping to improve this in ReservoirKit soon to improve the UX.

ximxim commented on src/utils/getCollections.ts:
```diff
@@ -0,0 +1,37 @@
+    contract: contracts,
+  })
+
+  const response = await fetch(url.href)
```
```suggestion
  const response = await fetch(url.href, { next: { revalidate: 0.5 * 60 } })
```

30s sounds good to me

cb-heimdall commented:
Review Error for ximxim @ 2023-08-18 17:02:09 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

pedromcunha commented on src/app/api/reservoir/[...slug]/route.ts:
```diff
@@ -0,0 +1,131 @@
+  { params }: { params: { slug: string | string[] } }
+) => {
+  const { body, method, headers: reqHeaders } = req
+  if (allowedDomains && allowedDomains.length > 0) {
+    let origin =
+      reqHeaders.get('origin') ||
+      reqHeaders.get('referer') ||
+      reqHeaders.get('host') ||
+      ''
+    if (origin) {
+      const hostMatches = origin.match(hostRegex)
+      origin = hostMatches && hostMatches[1] ? hostMatches[1] : origin
+    }
+    if (!origin.length || !allowedDomains.includes(origin)) {
+      return NextResponse.json(
+        { error: 'Access forbidden' },
+        {
+          status: 403,
+        }
+      )
+    }
+  }
```
Yep, exactly. You can configure allowed domains via the env variables, this just protects piggybacking off the reservoir proxy api I set up in the repo to hide the api key.

cb-heimdall commented:
Review Error for ximxim @ 2023-08-18 18:50:00 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

wilsoncusack commented:
will hold until monday I think

lauchness commented on src/components/MintButton/MintButton.tsx:
```diff
@@ -36,11 +43,15 @@ export const MintButton: FC<MintButtonProps> = ({ size, ...mintProps }) => {
 
   if ((mintProps.endDate && now >= mintProps.endDate) || isMintedOut) {
+    if (mintProps.address === '0x') {
+      return <ExternalLinkButton partner={mintProps.partnerName} />
```
This line will never execute.

The `MintButton` is only used in ternaries with ExternalDrop - or on the trending page with mint.fun drops.

In either case, the drop will never be an external drop at this line

wilsoncusack replied:
@lauchness what should the fallback be here, instead? Just "Mint ended/minted out" ? we can use the message from validate? cc @pedromcunha

pedromcunha replied:
Just to be clear, are you saying that the address will never be `0x` ? If that's the case then we can simply remove the check and ExternalLinkButton. The CollectButton requires a valid address, otherwise the modal will load infinitely.

lauchness reviewed (changes requested):
Preview deploy is broken for partner pages with Collect buttons: https://onchainsummer-xyz-git-fork-pedromcunha-feature-a9cca0-base-org.vercel.app/Coca-Cola

https://onchainsummer-xyz-git-fork-pedromcunha-feature-a9cca0-base-org.vercel.app/fwb

cb-heimdall commented:
Review Error for lauchness @ 2023-08-22 16:21:01 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

pedromcunha commented:
> Review Error for lauchness @ 2023-08-22 16:21:01 UTC User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

Odd, can't reproduce this locally, even with all env variables removed. Do you have any info on the error produced by visiting the page?

lauchness commented:
> > Review Error for lauchness @ 2023-08-22 16:21:01 UTC User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)
> 
> Odd, can't reproduce this locally, even with all env variables removed. Do you have any info on the error produced by visiting the page?

The only error is that something happened during the server components rendering
<img width="742" alt="Screen Shot 2023-08-22 at 3 08 36 PM" src="https://github.com/base-org/onchainsummer.xyz/assets/51837850/2a4d40a0-04f3-4c5e-8dd0-9a477b4b4b9a">

pedromcunha commented:
> > > Review Error for lauchness @ 2023-08-22 16:21:01 UTC User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)
> > 
> > 
> > Odd, can't reproduce this locally, even with all env variables removed. Do you have any info on the error produced by visiting the page?
> 
> The only error is that something happened during the server components rendering <img alt="Screen Shot 2023-08-22 at 3 08 36 PM" width="742" src="https://user-images.githubusercontent.com/51837850/262443614-2a4d40a0-04f3-4c5e-8dd0-9a477b4b4b9a.png">

Right, do you see anything else in the vercel logs?

pedromcunha commented:
> > > Review Error for lauchness @ 2023-08-22 16:21:01 UTC User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)
> > 
> > 
> > Odd, can't reproduce this locally, even with all env variables removed. Do you have any info on the error produced by visiting the page?
> 
> The only error is that something happened during the server components rendering <img alt="Screen Shot 2023-08-22 at 3 08 36 PM" width="742" src="https://user-images.githubusercontent.com/51837850/262443614-2a4d40a0-04f3-4c5e-8dd0-9a477b4b4b9a.png">

Fixed the crash, but seems like maybe the env variables are configured incorrectly on the preview link (this is because we haven't deployed to production yet. The reservoir base url (`NEXT_PUBLIC_RESERVOIR_API_URL`) is mapping to `https://onchainsummer.xyz/api/reservoir` which is not available in production yet. You can avoid configuring this all together and just configure it on production, it's only important for scaling.
